File: 96/spark -s51-81/s71.scala
Problem Scenario 71 . 
Write down a Spark script using Python, 
In which it read a file "Content.txt" (On hdfs) with following content. 
After that split each row as (key, value), where key is first word in line and entire line as value. 
Filter out the empty lines. 
And save this key value in "problem86" as Sequence file(On hdfs) 
Part 2 : Save as sequence file , where key as null and entire line as value. Read back the stored sequence files. 

Content.txt 
Hello this is HadoopExam.com 
This is QuickTechie.com 
Apache Spark Training 
This is Spark Learning Session 
Spark is faster than MapReduce 

val contentRDD = sc.textFile("spark2/Content.txt")

scala> contentRDD.take(10).foreach(println)
Hello this is HadoopExam.com
This is QuickTechie.com
Apache Spark Training
This is Spark Learning Session
Spark is faster than MapReduce

val contentRDDMap = contentRDD.map(rec => (rec.split(" ")(0), rec))
scala> contentRDDMap.take(10).foreach(println)
(Hello,Hello this is HadoopExam.com )
(This,This is QuickTechie.com )
(Apache,Apache Spark Training )
(This,This is Spark Learning Session )
(Spark,Spark is faster than MapReduce )

Save as sequence file , where key as null and entire line as value. Read back the stored sequence files.

contentRDDMap.saveAsSequenceFile("problem86")

[paslechoix@gw01 data]$ hdfs dfs -ls problem86
Found 3 items
-rw-r--r--   3 paslechoix hdfs          0 2018-02-09 20:24 problem86/_SUCCESS
-rw-r--r--   3 paslechoix hdfs        198 2018-02-09 20:24 problem86/part-00000
-rw-r--r--   3 paslechoix hdfs        169 2018-02-09 20:24 problem86/part-00001

ContentWithBlankLines.txt

Hello this is HadoopExam.com 
This is QuickTechie.com 

Apache Spark Training 
This is Spark Learning Session 

Spark is faster than MapReduce 

[paslechoix@gw01 data]$ vim ContentWithBlankLines.txt
[paslechoix@gw01 data]$ hdfs dfs -put ContentWithBlankLines.txt ContentWithBlankLines.txt
[paslechoix@gw01 data]$ hdfs dfs -ls ContentWithBlankLines.txt
-rw-r--r--   3 paslechoix hdfs        144 2018-02-09 20:33 ContentWithBlankLines.txt
[paslechoix@gw01 data]$ hdfs dfs -cat ContentWithBlankLines.txt

Hello this is HadoopExam.com
This is QuickTechie.com

Apache Spark Training
This is Spark Learning Session

Spark is faster than MapReduce
[paslechoix@gw01 data]$

scala> contentWithBlankLinesRDD.collect
res5: Array[String] = Array("", "Hello this is HadoopExam.com ", "This is QuickTechie.com ", "", "Apache Spark Training ", "This is Spark Learning Session ", "", Spark is faster than MapReduce)

scala> val contentWithBlankLinesRDDMap = contentWithBlankLinesRDD.map(rec =>(rec.split(" ")(0), rec))

scala> contentWithBlankLinesRDDMap.take(10).foreach(println)
(,)
(Hello,Hello this is HadoopExam.com )
(This,This is QuickTechie.com )
(,)
(Apache,Apache Spark Training )
(This,This is Spark Learning Session )
(,)
(Spark,Spark is faster than MapReduce)

scala> val contentFilteredRDD = contentWithBlankLinesRDDMap.filter(rec => rec._1.length>0)

(Hello,Hello this is HadoopExam.com )
(This,This is QuickTechie.com )
(Apache,Apache Spark Training )
(This,This is Spark Learning Session )
(Spark,Spark is faster than MapReduce)

File: 96/spark -s51-81/s71.py
Problem Scenario 71 . 
Write down a Spark script using Python, 
In which it read a file "Content.txt" (On hdfs) with following content. 
After that split each row as (key, value), where key is first word in line and entire line as value. 
Filter out the empty lines. 
And save this key value in "problem86" as Sequence file(On hdfs) 
Part 2 : Save as sequence file , where key as null and entire line as value. Read back the stored sequence files. 
Content.txt 
Hello this is HadoopExam.com 
This is QuickTechie.com 
Apache Spark Training 
This is Spark Learning Session 
Spark is faster than MapReduce 
=========================================================================== 

Solution : 

Data preparation:

[paslechoix@gw03 ~]$ cat p71.txt
Hello this is HadoopExam.com

This is QuickTechie.com
Apache Spark Training

This is Spark Learning Session
Spark is faster than MapReduce
[paslechoix@gw03 ~]$ hdfs dfs -put p71.txt 

[paslechoix@gw03 ~]$ hdfs dfs -cat p71.txt
Hello this is HadoopExam.com

This is QuickTechie.com
Apache Spark Training

This is Spark Learning Session
Spark is faster than MapReduce
[paslechoix@gw03 ~]$


Step 1 : 
# Import SparkContext and SparkConf 
from pyspark import SparkContext, SparkConf 
step 2 : 
#load data from hdfs 
p71RDD = sc.textFile("p71.txt") 
step 3 : 

>>> for line in nonempty_lines.collect(): print(line)

#filter out non-empty lines 
nonempty_lines = p71RDD.filter(lambda x: len(x) > O) 
Hello this is HadoopExam.com
This is QuickTechie.com
Apache Spark Training
This is Spark Learning Session
Spark is faster than MapReduce

Step 4: 
#Split line based on space (Remember : It is mandatory to convert is in tuple) 

words = nonempty_lines.map(lambda x: tuple(x.split(' ',1))) 

for w in words.collect(): print(w)
(u'Hello', u'this is HadoopExam.com ')
(u'This', u'is QuickTechie.com ')
(u'Apache', u'Spark Training ')
(u'This', u'is Spark Learning Session ')
(u'Spark', u'is faster than MapReduce ')

words.saveAsSequenceFile("problem86") 

Step 5: Check contents in directory problem86 
[paslechoix@gw03 ~]$ hdfs dfs -cat problem86_seq/*
SEQorg.apache.hadoop.io.Textorg.apache.hadoop.io.Text▒^▒T▒w▒kl|%▒}O+Hellothis is HadoopExam.com Thisis QuickTechie.com ApacheSpark Training SEQorg.apache.hadoop.io.Textorg.apache.hadoop.io.Text͏oBf▒▒P▒=▒▒▒ Thisis Spark Learning Session  Sparkis faster than MapReduce [paslechoix@gw03 ~]$


Step 6 : Reading back the sequence file data using spark. 
seqRDD = sc.sequenceFile(problem86_seq") 
(u'Hello', u'this is HadoopExam.com ')
(u'This', u'is QuickTechie.com ')
(u'Apache', u'Spark Training ')
(u'This', u'is Spark Learning Session ')
(u'Spark', u'is faster than MapReduce ')

